main.go

```diff
@@ -154,16 +172,74 @@ func main() {
 	}
 
 	go func() {
+
+		if !mgr.GetCache().WaitForCacheSync(context.TODO()) {
+			setupLog.Error(err, "Cache cannot start")
+			os.Exit(1)
+		}
+
+		MTLSconfig := mtls.NewMTLSconfig(mgr.GetClient(), operatorNamespace,
+			[]string{Config.Domain}, Config.TLSLocalhostEnabled)
+
+		tlsConfig, CACertChain, err := MTLSconfig.InitCertificates()
+		if err != nil {
+			setupLog.Error(err, "Cannot retrieve any MTLS configuration")
+			os.Exit(1)
+		}
+
+		// @TODO check here what to do with leftovers or if a new one is need to be created
+		err = MTLSconfig.CreateRegistrationClient()
+		if err != nil {
+			setupLog.Error(err, "Cannot create registration client certificate")
+			os.Exit(1)
+		}
+
+		opts := x509.VerifyOptions{
+			Roots:         tlsConfig.ClientCAs,
+			Intermediates: x509.NewCertPool(),
+		}
+
+		yggdrasilAPIHandler := yggdrasil.NewYggdrasilHandler(
+			edgeDeviceRepository,
+			edgeDeploymentRepository,
+			claimer,
+			initialDeviceNamespace,
+			mgr.GetEventRecorderFor("edgedeployment-controller"))
+
 		h, err := restapi.Handler(restapi.Config{
-			YggdrasilAPI: yggdrasil.NewYggdrasilHandler(edgeDeviceRepository, edgeDeploymentRepository, claimer, initialDeviceNamespace, mgr.GetEventRecorderFor("edgedeployment-controller")),
+			YggdrasilAPI: yggdrasilAPIHandler,
+			InnerMiddleware: func(h http.Handler) http.Handler {
+				// This is needed for one reason. Registration endpoint can be
+				// triggered with a certificate signed by the CA, but can be expired
+				// The main reason to allow expired certificates in this endpoint, it's
+				// to renew client certificates, and because some devices can be
+				// disconnected for days and does not have the option to renew it.
+				return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					authType := yggdrasilAPIHandler.SetAuthType(r)
+					if !mtls.VerifyRequest(r, authType, opts, CACertChain) {
```

Comment:
Will this be affected by exposing a service as a route and getting a different URL for yggdrasil server ?
 (by `oc expose svc k4e-operator-controller-manager -n k4e-operator-system`).

```
→ oc get routes -n k4e-operator-system k4e-operator-controller-manager
NAME                              HOST/PORT                                                                       PATH   SERVICES                          PORT   TERMINATION   WILDCARD
k4e-operator-controller-manager   k4e-operator-controller-manager-k4e-operator-system.apps.mycluster.redhat.com          k4e-operator-controller-manager   http                 None
```

Reply:
No, that's the part of the server, what we're checking here is the client certificates. So no an issue at all.